ios/sound.mm

```diff
@@ -24,209 +24,279 @@
-
-        for ( int iCh = 0; iCh < lNumInChanPlusAddChan; iCh++ )
-        {
-            int iSelCH, iSelAddCH;
-
-            GetSelCHAndAddCH ( iCh, lNumInChan, iSelCH, iSelAddCH );
-
-            if ( iSelAddCH >= 0 )
-            {
-                // for mixed channels, show both audio channel names to be mixed
-                channelInputName[iCh] =
-                    channelInputName[iSelCH] + " + " + channelInputName[iSelAddCH];
-            }
-        }
+        OSStatus err = AudioOutputUnitStart ( audioUnit );
+        checkStatus ( err );
     }
-    else
+    catch ( const CGenErr& generr )
     {
-        // regular case: no mixing input channels used
-        lNumInChanPlusAddChan = lNumInChan;
+        qDebug ( "Sound Start exception ...." );
     }
-*/
-    // everything is ok, return empty string for "no error" case
-    return "";
 }
 
-void CSound::SetLeftInputChannel  ( const int iNewChan )
+void CSound::Stop()
 {
-    /*// apply parameter after input parameter check
-    if ( ( iNewChan >= 0 ) && ( iNewChan < iNumInChanPlusAddChan ) )
+    try
+    {
+        OSStatus err = AudioOutputUnitStop ( audioUnit );
+        checkStatus ( err );
+    }
+    catch ( const CGenErr& generr )
     {
-        iSelInputLeftChannel = iNewChan;
+        qDebug ( "Sound Stop exception ...." );
     }
-     */
+    // call base class
+    CSoundBase::Stop();
 }
 
-void CSound::SetRightInputChannel ( const int iNewChan )
+void CSound::SetInputDeviceId ( int deviceid )
 {
- /* // apply parameter after input parameter check
-  if ( ( iNewChan >= 0 ) && ( iNewChan < lNumInChanPlusAddChan ) )
-  {
-      vSelectedInputChannels[1] = iNewChan;
-  }*/
-}
+    try
+    {
+        NSError* error      = nil;
+        bool     builtinmic = true;
 
-void CSound::SetLeftOutputChannel  ( const int iNewChan )
-{
-  /*// apply parameter after input parameter check
-  if ( ( iNewChan >= 0 ) && ( iNewChan < lNumOutChan ) )
-  {
-      vSelectedOutputChannels[0] = iNewChan;
-  }*/
-}
+        if ( deviceid == 0 )
+            builtinmic = false; // try external device
 
-void CSound::SetRightOutputChannel ( const int iNewChan )
-{
-  /*// apply parameter after input parameter check
-  if ( ( iNewChan >= 0 ) && ( iNewChan < lNumOutChan ) )
-  {
-      vSelectedOutputChannels[1] = iNewChan;
-  }*/
-}
+        AVAudioSession* sessionInstance = [AVAudioSession sharedInstance];
 
+        // assumming iOS only has max 2 inputs: 0 for builtin mic and 1 for external device
+        if ( builtinmic )
+        {
+            [sessionInstance setPreferredInput:sessionInstance.availableInputs[0] error:&error];
+        }
+        else
+        {
+            unsigned long lastInput = sessionInstance.availableInputs.count - 1;
+            [sessionInstance setPreferredInput:sessionInstance.availableInputs[lastInput] error:&error];
+        }
+    }
+    catch ( const CGenErr& generr )
+    {
+        qDebug ( "Sound dev change exception ...." );
```

Comment:
All these errors don't seem to be shown to the user. Shouldn't we output them as a warning message?

Reply:
using QMessageBox?

Reply:
Yes, to keep in Qt not native objective C.